.github/workflows/ci.yml

```diff
@@ -0,0 +1,68 @@
+name: CI
+
+on:
+  push:
+    # Sequence of patterns matched against refs/heads
+    branches:
+      - master
+  pull_request:
+    types:
+      - labeled
+      - unlabeled
+      - synchronize
+      - opened
+      - edited
+      - ready_for_review
+      - reopened
+      - unlocked
+  workflow_dispatch:
+  status: {}
+jobs:
+  # Build and run tests. Note that the valgrind tests crash.
+  build-and-test:
+    name: Build monocypher and run unit tests
+    strategy:
+      matrix:
+        platform: [ubuntu-latest] #macos-latest could also be here, but not free for Open Source
+    runs-on: ${{ matrix.platform }}
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Ubuntu apt-get update
+        run: sudo apt-get update
+
+      - name: Install libsodium
+        run: sudo apt-get install -y libsodium-dev
+
+      - name: Install clang and llvm
+        run: sudo apt-get install -y clang llvm
+
+      - name: Install valgrind
+        run: sudo apt-get install -y valgrind
+
+      - name: Make test vectors
+        run: |
+          pushd tests/gen
+          make clean && make
+          popd
+
+      - name: Run tests
+        run: make clean && make test
+
+      - name: Run speed tests
+        run: |
+          make clean && make speed
+
+      - name: Run code coverage tests
+        run: |
+          ./tests/coverage.sh
+
+      # This test fails with a crash. The set +e handles this for now, but the source of the crash
+      # should be found and fixed.
```

Comment:
Note that those tests work on my machine. On Github however, it would seem Valgrind does not work as it should:

```
vex amd64->IR: unhandled instruction bytes: 0x62 0xF1 0xFD 0x28 0x6F 0x35 0x9 0x5E 0x13 0x0
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=0 VEX.L=0 VEX.nVVVV=0x0 ESC=NONE
vex amd64->IR:   PFX.66=0 PFX.F2=0 PFX.F3=0
==3862== valgrind: Unrecognised instruction at address 0x15d28d.
==3862==    at 0x15D28D: vector_of_string (in /home/runner/work/Monocypher/Monocypher/test.out)
==3862==    by 0x15D764: next_input (in /home/runner/work/Monocypher/Monocypher/test.out)
==3862==    by 0x15B839: chacha20 (in /home/runner/work/Monocypher/Monocypher/test.out)
==3862==    by 0x15D8B0: vector_test (in /home/runner/work/Monocypher/Monocypher/test.out)
==3862==    by 0x159279: main (in /home/runner/work/Monocypher/Monocypher/test.out)
==3862== Your program just tried to execute an instruction that Valgrind
==3862== did not recognise.  There are two possible reasons for this.
==3862== 1. Your program has a bug and erroneously jumped to a non-code
==3862==    location.  If you are running Memcheck and you just saw a
==3862==    warning about a bad jump, it's probably your program's fault.
==3862== 2. The instruction is legitimate but Valgrind doesn't handle it,
==3862==    i.e. it's Valgrind's fault.  If you think this is the case or
==3862==    you are not sure, please let us know and we'll try to fix it.
==3862== Either way, Valgrind will now raise a SIGILL signal which will
==3862== probably kill your program.
```

I did not see a warning about a bad jump, so I assume this is Valgrind's fault. It would be nice to know which version of everything we're using here, so we may fix the problem.

Reply:
Valgrind on github actions is using 3.18. I am guessing your local machine is on 3.15. However, I think I have found a fix - installing valgrind via homebrew.

Reply:
Or, GCC is more recent on Github, and generates instructions Valgrind doesn't handle yet. Here's what I get on my machine:

```
$ uname -a
Linux old-grey 5.13.0-28-generic #31-Ubuntu SMP Thu Jan 13 17:41:06 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

$ gcc --version
gcc (Ubuntu 11.2.0-7ubuntu2) 11.2.0

$ valgrind --version
valgrind-3.17.0
```

If you have a quick fix, it would be nice if you could apply it in a third commit of your pull request. If it doesn't work that's too bad, but what we have now is still worth merging.

Reply:
My "homebrew fix" was very flakey. It worked once, then didn't work again (even rerunning the same commit.)
Still checking for a solution...

Reply:
I did "resolve" in this action: https://github.com/cfogelklou/Monocypher/actions/runs/1955574996
However, the fix was to build with CMake & Clang instead of GCC.
That same merge also adds cross-platform building testing on windows and macOS :)

Reply:
Hmm, There’s probably a way to build with Clang and regular Make, then run Valgrind on that. We should test that.

Just one more thing before I merge: is there a way to add comments at the beginning of the GitHub actions file? If so, it would be nice if you could add the licence info like there is in pretty much all other files (just copy one at random, and adjust the author name & year).

Reply:
Sure. I'll add the comments and try building with Clang. I'll do that and then update this PR. Hold tight.

Reply:
Maybe I'm thinking too far ahead, but maybe there's a weird interaction between valgrind and gcc's codegen in those particular versions.

What about replacing `ubuntu-latest` with some other kind of Linux base image?

(Evidently, all other tests run normally, so it's not like there's a direct codegen bug in gcc itself, since the normal tests with illegal instructions should fail equally.)

Reply:
Hi, I found the issue. It doesn't matter if clang or gcc is used, the culprit is the flag "-march=native", regardless of optimization level.

I think it is a good idea to leave failing tests failing, so the failing tests are left in, but ignored. But the workflow finishes with passing valgrind runs, so any subsequent failures introduced by anyone will cause those tests to fail.

You can check the result here: 
[https://github.com/cfogelklou/Monocypher/runs/5506724695?check_suite_focus=true
](https://github.com/cfogelklou/Monocypher/runs/5506724695?check_suite_focus=true
)

Reply:
Looks good to me, merging. Thanks.

If the only culprit is `march=native`, I think we can safely ignore the problem, especially if we have establish that it's Valgrind's fault for not handling the instruction — the program works after all.